### PR TITLE
Required control

### PR DESCRIPTION
**Pull request by AhmetTanakol**
- Required field control move to `mapStateToControlProps` function
- Required field warning is set in control classes

### CODE REVIEW

**Comment by edgarmueller** [comment]
@eneufeld Ping me, if you have any questions about the requirements.

File: src/renderers/controls/date.control.tsx [hunk 1]
```diff
@@ -21,14 +21,14 @@ export const dateControlTester: RankedTester = rankWith(2, and(
...
-    const { classNames, id, visible, enabled, errors, label, uischema } = this.props;
+    const { classNames, id, visible, enabled, errors, label, uischema, required } = this.props;
     const isValid = errors.length === 0;
     const divClassNames = 'validation' + (isValid ? '' : ' validation_error');
 
     return (
       <div className={classNames.wrapper}>
         <label htmlFor={id} className={classNames.label} data-error={errors}>
-          {label}
+          {required ? label + '*' : label}
```
Comment (eneufeld) [inline src/renderers/controls/date.control.tsx]: please extract this to a util to remove code duplication

File: src/renderers/label.util.ts [hunk 2]
```diff
@@ -67,10 +55,6 @@ const isRequired = (schema: JsonSchema, schemaPath: string): boolean => {
  */
 export const getElementLabelObject =
```
Comment (eneufeld) [inline src/renderers/label.util.ts]: This method is now obsolete, please remove it and export getLabelObject instead. You will need to update all the imports of course.

File: src/renderers/controls/date.control.tsx (hunk 1 above)
Comment (eneufeld) [reply src/renderers/controls/date.control.tsx]: a good place would be the renderer.util

Comment (AhmetTanakol) [reply src/renderers/controls/date.control.tsx]: should I remove `required` field from `props`

Comment (AhmetTanakol) [reply src/renderers/controls/date.control.tsx]: So you want me to create a function in renderer.util, so controllers will call this function with the `required` parameter. Am I correct?

Comment (edgarmueller) [reply src/renderers/controls/date.control.tsx]: @AhmetTanakol Yes, e.g. think about what would be necessary if you change the asterisk to a different character: You'd need to change the line in each renderer, which is cumbersome.

**Comment by AhmetTanakol** [comment]
I first pushed from wrong local branch but then I checkout the correct one and I used cherry-pick and updated the correct local branch and pushed again. I hope it is fixed now

File: src/renderers/label.util.ts [hunk 3]
```diff
@@ -26,7 +25,14 @@ const deriveLabel = (controlElement: ControlElement): string => {
...
 
-const getLabelObject = (withLabel: ControlElement): ILabelObject => {
+/**
+ * Return a label object based on the given JSON schema and control element.
+ * @param {JsonSchema} schema the JSON schema that the given control element is referring to
+ * @param {ControlElement} controlElement the UI schema to obtain a label object for
+ * @returns {ILabelObject}
+ */
+
+export const getLabelObject = (withLabel: ControlElement): ILabelObject => {
```
Comment (eneufeld) [inline src/renderers/label.util.ts]: The JavaDoc doesn't fit the method

File: src/renderers/renderer.util.tsx [hunk 4]
```diff
@@ -121,16 +122,34 @@ export const registerStartupRenderer = (tester: RankedTester, renderer: any) =>
...
+    const nextHigherSchemaSegments = pathSegments.slice(0, pathSegments.length - 2);
+    const nextHigherSchemaPath = nextHigherSchemaSegments.join('/');
+    const nextHigherSchema = resolveSchema(schema, nextHigherSchemaPath);
+
+    return nextHigherSchema !== undefined
+        && nextHigherSchema.required !== undefined
+        && nextHigherSchema.required.indexOf(lastSegment) !== -1;
+};
+
+export const setLabelField = (label: string, required: boolean): string => {
```
Comment (eneufeld) [inline src/renderers/renderer.util.tsx]: what does this method set? Doesn't it only return something?

Comment (AhmetTanakol) [reply src/renderers/renderer.util.tsx]: Yes, at this point I am not sure where to define label setting. Or is it ok to use a method that returns the edited version of label in the control? Maybe it has to be set in the control's constructor. Do you have any advice?

Comment (eneufeld) [reply src/renderers/renderer.util.tsx]: I would simply rename the method, to describe what it does. Something like `computeLabel`.

**Pull request merged**